game-core/src/main/java/games/strategy/engine/data/MoveDescription.java

```diff
@@ -0,0 +1,77 @@
+package games.strategy.engine.data;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.HashMultiset;
+import com.google.common.collect.Maps;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.Getter;
+
+/** Describes an action that moves one or more units along a specific route. */
+public class MoveDescription extends AbstractMoveDescription {
+  private static final long serialVersionUID = 2199608152808948043L;
+  @Getter private final Route route;
+  @Getter private final Map<Unit, Unit> unitsToTransports;
+  @Getter private final Map<Unit, Collection<Unit>> dependentUnits;
+
+  public MoveDescription(
+      final Collection<Unit> units,
+      final Route route,
+      final Map<Unit, Unit> unitsToTransports,
+      final Map<Unit, Collection<Unit>> dependentUnits) {
+    super(Collections.unmodifiableCollection(units));
+    this.route = Preconditions.checkNotNull(route);
+    this.unitsToTransports = Collections.unmodifiableMap(unitsToTransports);
+    this.dependentUnits =
+        Collections.unmodifiableMap(
+            dependentUnits.entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> Set.copyOf(e.getValue()))));
+  }
+
+  public MoveDescription(
+      final Collection<Unit> units, final Route route, final Map<Unit, Unit> unitsToTransports) {
+    this(units, route, unitsToTransports, Map.of());
+  }
+
+  public MoveDescription(final Collection<Unit> units, final Route route) {
+    this(units, route, Map.of(), Map.of());
+  }
+
+  @Override
+  public final boolean equals(final Object obj) {
+    if (this == obj) {
+      return true;
+    } else if (!(obj instanceof MoveDescription)) {
+      return false;
+    }
+
+    final MoveDescription other = (MoveDescription) obj;
+    return getClass().equals(other.getClass())
+        && route.equals(other.route)
+        && Maps.difference(unitsToTransports, other.unitsToTransports).areEqual()
+        && Maps.difference(dependentUnits, other.dependentUnits).areEqual()
+        && collectionsAreEqual(getUnits(), other.getUnits());
+  }
+
+  @Override
+  public final int hashCode() {
+    return Objects.hash(
+        HashMultiset.create(getUnits()), route, new HashMap<>(unitsToTransports), dependentUnits);
+  }
+
+  @SuppressWarnings("UndefinedEquals")
+  private static boolean collectionsAreEqual(final Collection<Unit> a, final Collection<Unit> b) {
+    // https://stackoverflow.com/questions/1565214/is-there-a-way-to-check-if-two-collections-contain-the-same-elements-independen
+    return HashMultiset.create(a).equals(HashMultiset.create(b));
+  }
+
+  @Override
+  public String toString() {
+    return "Move message route:" + route + " units:" + getUnits();
```

Comment:
I'm a bit confused by "Move message", what does message mean in this context?
Have you considered using lombok `@ToString` to avoid a custom toString implementation?

Same thing for `hashCode` and `equals`, you can use `@EqualsAndHashcode` to avoid writing those on your own.

Reply:
I'm not sure. This class is being moved from games.strategy.triplea.delegate.data package and I didn't change the implementation of this method.

Let me know if you prefer I change it to use @ToString assuming nothing relies on this message.

For equals & hashCode, I'm not sure the defaults would work since the collections may have different underlying implementations. I just care about the contents, not what their underlying types are.

Reply:
I'd hope nothing relies on this message, it seems here for debug.
W.r.t equals, we can represent the internal collection as `Set` to indicate we care that ordering does not matter.

Reply:
Regardless, since we're doing a file move mostly, this kind of update would be best left for follow-up.